### ISSUE TITLE
Creating archive with Xcode 9.1 fails

### ISSUE THREAD

**Issue by oemera** [flutter/flutter#13070]
We are trying to create a IPA file with Xcode 9 commandline tools and are running into issues. If we do the same procedure with Xcode 8.3.3 it works fine so we blieve the problem is tied to Xcode 9.1.

## Steps to Reproduce

1. `flutter create -i swift --org com.example.xcode_error .`
2. Open Xcode and set `Signing` to `Automatically manage signing`
3. Set device to`Generic iOS Device`
4. Choose `Product` -> `Archive`
5. After successfully archiving choose app in `Organizer`
6. `Export...` the app as `Enterprise` deployment

After a couple seconds you will see this error:

```
ipatool failed with an exception: #<TypeError: no implicit conversion of nil into Array>
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2769:in `&'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2769:in `bestDeviceTraitsForBundle'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:1210:in `block in makeDeviceTraitsToDeviceTypesMapForDeviceIdentifiers'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:1195:in `each'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:1195:in `makeDeviceTraitsToDeviceTypesMapForDeviceIdentifiers'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2466:in `block in ProcessIPA'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2444:in `each'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2444:in `ProcessIPA'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2940:in `<main>'
```

## Logs

**Xcode ipatool logs:**
[IDEDistribution.verbose.log](https://github.com/flutter/flutter/files/1483364/IDEDistribution.verbose.log)

## Flutter Doctor

```
[✓] Flutter (on Mac OS X 10.12.6 16G29, locale en-DE, channel master)
    • Flutter at /Users/oemeravci/Documents/Development/XcodeWorkspace/Grandcentrix/sample_flutter/flutter
    • Framework revision d5d2cdfeef (12 hours ago), 2017-11-16 22:53:22 -0800
    • Engine revision 25912b8482
    • Tools Dart version 1.25.0-dev.11.0

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
      If Android SDK has been installed to a custom location, set $ANDROID_HOME to that location.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.1, Build version 9B55
    • ios-deploy 1.9.2
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b05)

[-] IntelliJ IDEA Community Edition (version 2017.2.4)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.io/intellij-setup/#installing-the-plugins

[✓] Connected devices
    • None
```

**Comment by cbracken** [comment]
Interesting, this looks potentially like a bug in `ipatool`, which is a Ruby script that's part of Xcode. Out of curiosity, does Xcode 9.2 resolve this issue for you?

I'll give a go at reproducing this today.

**Comment by passsy** [comment]
I'm not able to export an app as Enterprise deployment. Tested with Xcode 9.2 and flutter on `master` 77872cc5601a168f3644689b966b45223d90efc2 and `alpha` 8f65fec5f5f7d7afbb0965f4a44bdb330a28fb19.

New: Xcode tells me to first build the app in release mode.
```
========================================================================
ERROR: Flutter archive builds must be run in Release mode.

To correct, run:
flutter build ios --release

then re-run Archive from Xcode.
========================================================================
```

After doing so I get the same error:

```
ipatool failed with an exception: #<TypeError: no implicit conversion of nil into Array>
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2767:in `&'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2767:in `bestDeviceTraitsForBundle'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:1209:in `block in makeDeviceTraitsToDeviceTypesMapForDeviceIdentifiers'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:1194:in `each'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:1194:in `makeDeviceTraitsToDeviceTypesMapForDeviceIdentifiers'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2464:in `block in ProcessIPA'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2442:in `each'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2442:in `ProcessIPA'
    /Applications/Xcode.app/Contents/Developer/usr/bin/ipatool:2938:in `<main>'
```

![screen shot 2018-01-10 at 19 51 26](https://user-images.githubusercontent.com/1096485/34789729-af4102a2-f63f-11e7-8009-1e1a44551fac.png)

**Comment by pedia** [comment]
Me too. XCode behavior
I changed:
...
5. 
6. Validate...
7. Next
8. then Export

Succeeded.

**Comment by oemera** [comment]
@pedia thank you for pointing out that it works with `Validate` -> `Export`. This works for me too.

Unfortunately this workaround doesn't work with our Enterprise Developer Account because `Validate`  does validate the app with `iTunes Connect` and Enterprise Developer Accounts don't have access to iTunes Connect.

**Comment by kamal333k** [comment]
Hi guys, I'm facing this similar issue while creating an ipa. 
I tried to export after validating but on validating it asks me for app store provisioning profile.

I'm not trying to publish my application on app store. I want to create an ipa in adhoc environment. 
Can anyone help me on this?

**Comment by ben-wood** [comment]
This error must have something to do with App Thinning, If I click next quickly before the error comes up while it is checking for App Thinning then it goes right on with the export.

**Comment by pedia** [comment]
I prefer [fastlane](https://fastlane.tools) to finish the work of XCode, like:
adhoc, enterprise, upload to testflight, publish to appstore...

**Comment by xster** [comment]
Haven't tried to reproduce this yet but in the meantime, wrt fastlane, here's an example of our Gallery app being built, signed and pushed to iTunes Connect programmatically
https://github.com/flutter/flutter/blob/master/dev/bots/travis_script.sh#L21

Here's the actual Fastfile https://github.com/flutter/flutter/blob/master/examples/flutter_gallery/ios/fastlane/Fastfile though the whole thing could be a lot simpler if building locally instead of an untrusted CI system.

**Comment by sw-tt-jigneshkumarchitaliya** [comment]
Please change validate build product release Yes to No. For more detail please check the screenshot.
![runner](https://user-images.githubusercontent.com/25667876/37084150-c0551502-2217-11e8-920e-3b9c924880fe.png)

**Comment by fralways** [comment]
Still happening with the export option in xcode organizer - tho I get this error for Ad-Hoc option (didn't try enterprise). Clicking fast on Next button is a workaround and I was able to successfully distribute  generated ipa. 
(Changing validate build product release from Yes to No doesn't help)

**Comment by alescdb** [comment]
I was having this issue too with XCode 9.2.
I've update to 9.3 this morning and it's resolved !

**Comment by cbracken** [comment]
@alescdb thanks for this info.

For anyone that still has this issue, can you verify if you have a custom install of ruby on your machine? From the command-line, check the output of `which ruby`. If it's not `/usr/bin/ruby`, what happens if you adjust your `PATH` environment variable (e.g. in .bashrc or .profile) so that you're picking up `/usr/bin/ruby` by default?

Other potentially useful data: what is the output of `xcrun ipatool --help`? specifically the bits before the usage information.

**Comment by palmerc** [comment]
The problem is the version of ipatool in Xcode 9.2 cannot handle a Framework where the plist does not contain a UIDeviceFamily key-value pair. You can just manually add UIDeviceFamily = 1,2 to the framework plist and the problem should go away.

**Comment by zoechi** [comment]
Related to #12774

**Comment by Chetan8488** [comment]
Not's working Xcode 9.2 
Please update Xcode 10.1 
Then working fine...

**Comment by jmagman** [comment]
Looks like this ipatool bug was resolved with Xcode 9.3, which was released more than a year ago.  Closing.
